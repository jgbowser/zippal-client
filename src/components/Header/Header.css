@import url('https://fonts.googleapis.com/css2?family=Romanesco&family=The+Girl+Next+Door&display=swap');
header {
  background-color: rgba(255, 255, 255, 0.5);
  text-align: center;
  display: flex;
  width: 100%;
  align-items: center;
  flex-direction: row;
  justify-content: space-between;
  min-height: 12vmax;
  max-height: 12vmin;
  margin: 0;
  padding: 0;
}
.header_text,
h1 {
  display: flex;
  align-items: center;
}

img.quill {
  width: 80px;
}
.header_link {
  display: flex;
}
.nav_container {
  display: flex;
  flex-direction: column-reverse;
}
header a {
  font-size: 10vmin;
  font-family: 'Romanesco', cursive;
  color: black;
  text-decoration: none;
}

nav {
  display: flex;
}

nav a {
  margin: 0 20px;
  font-family: 'The Girl Next Door', cursive;
  text-decoration: none;
  font-size: 3vmax;
  color: black;
}

nav a:hover {
  color: #9c27ca;
}

@media all and (max-width: 360px) {
  nav {
    flex-direction: column;
    justify-content: space-evenly;
  }
}
