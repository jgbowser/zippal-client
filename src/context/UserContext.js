import React, {Component} from 'react'
import AuthApiService from '../services/auth-api-service'
import UserService from '../services/user-api-service'
import TokenService from '../services/token-service'
import IdleService from '../services/idle-service'

const UserContext = React.createContext({
  user:{},
  profileInfo:{},
  error:null,
  setError:() => { },
  clearError: () => { },
  setUser: () => { },
  getProfile: () => { },
  updateProfile: () => { },
  processLogin: () => { },
  processLogout: () => { },
})

export default UserContext

export class UserProvider extends Component {
  constructor(props){
    super(props)
    const state = {user: {}, error:null, profileInfo: {}}

    const jwtPayload = TokenService.parseAuthToken()
    

    if(jwtPayload){
      const profileData = this.getProfile()

      state.user = {
        id: jwtPayload.id,
        display_name: jwtPayload.display_name,
        username: jwtPayload.sub,
      }

      if(profileData)
        state.profileInfo = {
          bio: profileData.bio,
          display_name: profileData.display_name,
          fa_icon: profileData.fa_icon,
          location: profileData.location
      }
    }
      

    this.state = state;
    IdleService.setIdleCallback(this.logoutBecauseIdle)
  }

  componentDidMount(){
    if(TokenService.hasAuthToken()){
      IdleService.registerIdleTimerResets()
      TokenService.queueCallbackBeforeExpiry(() => {
        this.fetchRefreshToken()
      })
    }
  }

  componentWillUnmount() {
    IdleService.unRegisterIdleResets()
    TokenService.clearCallBackBeforeExpiry()
  }

  setError = error => {
    console.error(error)
    this.setState({error})
  }

  clearError = () => {
    this.setState({error: null})
  }
  
  setUser = user => {
    this.setState({user})
  }

  getProfile = async () => {
    const profileInfo = await UserService.getUserProfile()
    this.setState({profileInfo})
  }

  updateProfile = (profileInfo) => {
    this.setState({profileInfo})
  }

  processLogin = authToken => {
    TokenService.saveAuthToken(authToken)

    const jwtPayload = TokenService.parseAuthToken()
    this.setUser({
      id: jwtPayload.id,
      display_name: jwtPayload.display_name,
      username: jwtPayload.sub,
    })
    this.getProfile()
    IdleService.registerIdleTimerResets()
    TokenService.queueCallbackBeforeExpiry(() => {
      this.fetchRefreshToken()
    })
  }

  processLogout = () => {
    TokenService.clearAuthToken()
    TokenService.clearCallBackBeforeExpiry()
    IdleService.unRegisterIdleResets()
    this.setUser({})
    this.setState({profileInfo: {}})
  }

  logoutBecauseIdle = () => {
    TokenService.clearAuthToken()
    TokenService.clearCallBackBeforeExpiry()
    IdleService.unRegisterIdleResets()
    this.setUser({idle:true})
  }

  fetchRefreshToken = () => {
    AuthApiService.refreshToken()
      .then(res => {
        TokenService.saveAuthToken(res.authToken)
        TokenService.queueCallbackBeforeExpiry(() => {
          this.fetchRefreshToken()
        })
      })
      .catch(err => {
        this.setError(err)
      })
  }

  render() {
    const value = {
      user: this.state.user,
      profileInfo: this.state.profileInfo,
      error: this.state.error,
      setError: this.setError,
      clearError: this.clearError,
      setUser: this.setUser,
      getProfile: this.getProfile,
      updateProfile: this.updateProfile,
      processLogin: this.processLogin,
      processLogout: this.processLogout
    }
    return (
      <UserContext.Provider value={value}>
        {this.props.children}
      </UserContext.Provider>
    )
  }
}