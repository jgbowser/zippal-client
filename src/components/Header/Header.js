import React, {Component} from 'react'
import {Link} from 'react-router-dom'
import TokenService from '../../services/token-service.js'
import UserContext from '../../context/UserContext'
import quill from '../../images/quill.png'
import './Header.css'

class Header extends Component {
  static contextType = UserContext

  handleLogoutClick = () => {
    this.context.processLogout()
  }

  renderLogoutLink(){
    return (
      <div className='logged_in'>
      <div className='nav_container'>
        <nav className='nav'>
          <Link 
          onClick={this.handleLogoutClick} 
          to='/login'>
          Logout
          </Link>
          {' ' }
          <Link to='/profile'>Profile</Link>
        </nav>
      </div>
      </div>
    )
  }

  renderLoginLink(){
    return (
      <div className='nav_container'>
      <nav className='nav'>
        <Link to='/login'>Login</Link>
        {' ' }
        <Link to='/registration'>Sign up</Link>
      </nav>
      </div>
    )
  }

  render() {
    return (
      <>
         <h1 className="header_text">
            <div className='header_image'>
              {TokenService.hasAuthToken()
              ? <Link className='header_link' to='/dashboard'><img className='quill' alt='zip pal' src={quill}/></Link>
              : <Link className='header_link' to='/'><img className='quill'  alt='zip pal' src={quill}/></Link> }
            </div>
            <div className='title'>
              Zip Pal
            </div>
          </h1>
        {TokenService.hasAuthToken()
        ? this.renderLogoutLink()
        : this.renderLoginLink()}
      </>
    );
  }
}

export default Header