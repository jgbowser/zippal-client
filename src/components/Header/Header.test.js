import React from 'react'
import ReactDOM from 'react-dom'
import { BrowserRouter } from 'react-router-dom'
import UserContext from '../../context/UserContext' //Used to process logout
import Header from './Header'

describe('Header Component', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render(<BrowserRouter><Header /></BrowserRouter>, div)
    ReactDOM.unmountComponentAtNode(div)
  })
})