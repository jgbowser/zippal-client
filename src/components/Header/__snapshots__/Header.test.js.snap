// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component takes a snapshot of the header component 1`] = `
Array [
  <div
    className="header_text"
  >
    <div
      className="header_image"
    >
      <a
        className="header_link"
        href="/"
        onClick={[Function]}
      >
        <img
          alt="zip pal"
          className="quill"
          src="quill.png"
        />
      </a>
    </div>
    <h1
      className="title"
    >
      Zip Pal
    </h1>
  </div>,
  <div
    className="nav_container"
  >
    <nav
      className="nav"
    >
      <a
        href="/login"
        onClick={[Function]}
      >
        Login
      </a>
       
      <a
        href="/registration"
        onClick={[Function]}
      >
        Sign up
      </a>
    </nav>
  </div>,
]
`;
